### Pattara Udomsombutrattana

### 1.	My partner and I have recently rediscovered a love for Lego™. Not the children’s toy, but the really complex Lego with thousands of pieces, Star Wars themes, and fancy architecture (OK – fine – it’s still a children’s toy). Sometimes, we build Lego together, other times I build on my own. I would not admit this to her (at least, not without sufficient evidence), but I am pretty sure I am faster at building Lego alone than we are together. I have been collecting data on the time it takes to build the set, the number of pieces, the difficulty of the set, the number of differently coloured pieces, and who built it. 

In [8]:
#Import the required libraries
import pandas as pd
import numpy as np
import os.path as osp
import statsmodels.api as sm
from statsmodels.formula.api import ols
from sklearn.linear_model import LinearRegression

#Import Data from Excel file
data_path = osp.join(osp.curdir,'MFIT860_Exam_Data.xlsx')
Lego_data = pd.read_excel(data_path,sheet_name='Lego')
Lego_data.head()

Time_S  Num_Pieces  Num_Colour  Difficulty         Builder
0    7552         635          11           6              Me
1    7674         564           9           5              Me
2   11143         932           9           5  Me and Partner
3    9022         805           9           8              Me
4    7241         627          11           4  Me and Partner

In [9]:
#We create Builder as dummy variable
Builder_dummy= pd.get_dummies(Lego_data['Builder']) 
Lego_data= pd.concat([Lego_data,Builder_dummy],axis=1)
Lego_data= Lego_data.drop('Builder',axis=1)
Lego_data.head()

Time_S  Num_Pieces  Num_Colour  Difficulty  Me  Me and Partner
0    7552         635          11           6   1               0
1    7674         564           9           5   1               0
2   11143         932           9           5   0               1
3    9022         805           9           8   1               0
4    7241         627          11           4   0               1

#### a.	I believe that I am faster at building Lego than my partner and I are together in relation to the number of pieces.  Conduct the most powerful test you can to see if this is the case.  Briefly explain the test and your results (i.e. state the null, alternative, the p-value of the result and what it means.) 2 marks

In [39]:
#Build our model 
Lego_model = ols('Time_S ~ Num_Pieces + Num_Colour + Difficulty + Me',Lego_data).fit()
Lego_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Time_S   R-squared:                       0.946
Model:                            OLS   Adj. R-squared:                  0.944
Method:                 Least Squares   F-statistic:                     417.8
Date:                Thu, 06 Apr 2023   Prob (F-statistic):           2.34e-59
Time:                        19:34:07   Log-Likelihood:                -762.02
No. Observations:                 100   AIC:                             1534.
Df Residuals:                      95   BIC:                             1547.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    797.0327    368.650      2.162      0.033      65.170    1528.895
Num_Pieces    10.0295      0.246     40.735      0.000       9.541      10.518
Num_Colour    87.3928     25.931      3.370      0.001      35.913     138.873
Difficulty    12.8822     15.811      0.815      0.417     -18.507      44.271
Me          -131.2518    107.162     -1.225      0.224    -343.995      81.491
==============================================================================
Omnibus:                      238.986   Durbin-Watson:                   1.965
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               12.891
Skew:                          -0.452   Prob(JB):                      0.00159
Kurtosis:                       1.492   Cond. No.                     6.63e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.63e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [44]:
Lego_hypothesis1 = '(Num_Pieces=0, Me=0)'
print(Lego_model.wald_test(Lego_hypothesis1))

<F test: F=array([[835.62672012]]), p=5.091582264553673e-61, df_denom=95, df_num=2>


### Answer
If Model : Time_S = B0 + B1Num_Pieces + B2Num_Colour+ B3Difficulty 	+ B4Me build

Firstly set business problem : I can build lego faster than my partner and I are together if  data show that time_S is relate to number of pieces
H0 : B1 = B4 = 0
H1 : at least one ≠ 0

P-value is tiny at 5.091582264553673e-61 which is less than alpha at 0.05. Therefore we reject null hypothesis. This indicate that there is correlation between number of pieces and I build lego to explain Time_S.

#### b.	Now conduct a Chow test to determine if us building together and me by myself are different at all.  Briefly explain these results, particularly in light of your test above. 3 marks

In [45]:
#Create new column 
Lego_data['C1'] = Lego_data['Num_Pieces'] * Lego_data['Me']
Lego_data['C2'] = Lego_data['Num_Colour'] * Lego_data['Me']
Lego_data['C3'] = Lego_data['Difficulty'] * Lego_data['Me']
Lego_data.head()

Time_S  Num_Pieces  Num_Colour  Difficulty  Me  Me and Partner  C0  \
0    7552         635          11           6   1               0   1   
1    7674         564           9           5   1               0   1   
2   11143         932           9           5   0               1   1   
3    9022         805           9           8   1               0   1   
4    7241         627          11           4   0               1   1   

   my partner   C1  C2  C3  
0           1  635  11   6  
1           1  564   9   5  
2           1    0   0   0  
3           1  805   9   8  
4           1    0   0   0

In [48]:
#Build the model
Lego_modelb = ols('Time_S ~ Num_Pieces + Num_Colour + Difficulty + Me + C1 + C2 + C3',Lego_data).fit()
Lego_modelb.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Time_S   R-squared:                       0.947
Model:                            OLS   Adj. R-squared:                  0.943
Method:                 Least Squares   F-statistic:                     233.7
Date:                Thu, 06 Apr 2023   Prob (F-statistic):           1.05e-55
Time:                        19:58:30   Log-Likelihood:                -761.51
No. Observations:                 100   AIC:                             1539.
Df Residuals:                      92   BIC:                             1560.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    551.8089    691.409      0.798      0.427    -821.388    1925.006
Num_Pieces    10.3291      0.545     18.946      0.000       9.246      11.412
Num_Colour    72.8729     42.225      1.726      0.088     -10.990     156.736
Difficulty    34.5649     29.936      1.155      0.251     -24.890      94.020
Me           151.0021    822.787      0.184      0.855   -1483.125    1785.129
C1            -0.3608      0.613     -0.588      0.558      -1.579       0.858
C2            19.9175     54.444      0.366      0.715     -88.213     128.049
C3           -28.8027     35.624     -0.809      0.421     -99.555      41.949
==============================================================================
Omnibus:                      224.475   Durbin-Watson:                   1.941
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               12.165
Skew:                          -0.407   Prob(JB):                      0.00228
Kurtosis:                       1.497   Cond. No.                     2.25e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.25e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [49]:
#Now run the chow test which is just a specific joint hypothesis test!
hypothesisb = '(Me = 0, C1 = 0, C2 = 0, C3 = 0)'
print(Lego_modelb.wald_test(hypothesisb))

<F test: F=array([[0.60464627]]), p=0.6602622377862143, df_denom=92, df_num=4>


D:\New Program\lib\site-packages\statsmodels\base\model.py:1889: FutureWarning: The behavior of wald_test will change after 0.14 to returning scalar test statistic values. To get the future behavior now, set scalar to True. To silence this message while retaining the legacy behavior, set scalar to False.
  warnings.warn(


### Answer
P-value is at 0.6602622377862143 which is greater than alpha at 0.05.  we cannot reject the null hypothesis. This indicate that there is no different between building together and me by myself 


#### c.	Suppose the last 30 observations my partner actually built on her own – extend the chow test logic to test for the difference between any of the three scenarios (i.e., me alone, my partner alone, us together).  Explain your results. 3 marks

In [52]:
#Create new column 
Lego_data['C0'] = np.where(Lego_data['Time_S'].isin(Lego_data['Time_S'].tail(30)), 0, 1)
Lego_data['C4'] = Lego_data['Num_Pieces'] * Lego_data['C0']
Lego_data['C5'] = Lego_data['Num_Colour'] * Lego_data['C0']
Lego_data['C6'] = Lego_data['Difficulty'] * Lego_data['C0']

Lego_data[65:75]

Time_S  Num_Pieces  Num_Colour  Difficulty  Me  Me and Partner  C0  \
65    8829         794           8           2   1               0   1   
66    8364         607          10           1   1               0   1   
67   10894         877          11           7   0               1   0   
68   11234         900          11           5   0               1   1   
69   13847        1179          11           5   0               1   1   
70    7298         490          13           1   0               1   0   
71   13193        1132           8           1   0               1   0   
72   10609         840          12           6   1               0   0   
73    7510         625          12           3   1               0   0   
74    9093         809          11          10   0               1   0   

    my partner   C1  C2  C3    C4  C5  C6  
65           1  794   8   2   794   8   2  
66           1  607  10   1   607  10   1  
67           0    0   0   0     0   0   0  
68           1    0   0   0   900  11   5  
69           1    0   0   0  1179  11   5  
70           0    0   0   0     0   0   0  
71           0    0   0   0     0   0   0  
72           0  840  12   6     0   0   0  
73           0  625  12   3     0   0   0  
74           0    0   0   0     0   0   0

In [57]:
#Build the model
Lego_modelc = ols('Time_S ~ Num_Pieces + Num_Colour + Difficulty + Me + C0 + C1 + C2 + C3 + C4 + C5 + C6',Lego_data).fit()
Lego_modelc.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Time_S   R-squared:                       0.949
Model:                            OLS   Adj. R-squared:                  0.943
Method:                 Least Squares   F-statistic:                     149.2
Date:                Thu, 06 Apr 2023   Prob (F-statistic):           6.13e-52
Time:                        20:13:06   Log-Likelihood:                -759.25
No. Observations:                 100   AIC:                             1543.
Df Residuals:                      88   BIC:                             1574.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -485.3848    954.888     -0.508      0.613   -2383.024    1412.254
Num_Pieces    11.0615      0.692     15.984      0.000       9.686      12.437
Num_Colour   107.3334     59.128      1.815      0.073     -10.170     224.837
Difficulty    31.8797     35.972      0.886      0.378     -39.607     103.367
Me           -11.9028    838.348     -0.014      0.989   -1677.943    1654.137
C0          1437.4069    919.414      1.563      0.122    -389.735    3264.549
C1            -0.2251      0.626     -0.359      0.720      -1.470       1.019
C2            20.5887     55.307      0.372      0.711     -89.323     130.500
C3           -23.3596     37.263     -0.627      0.532     -97.412      50.693
C4            -1.0905      0.610     -1.788      0.077      -2.302       0.121
C5           -36.6721     62.098     -0.591      0.556    -160.078      86.734
C6            -2.0575     34.397     -0.060      0.952     -70.413      66.299
==============================================================================
Omnibus:                       67.433   Durbin-Watson:                   1.929
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               10.124
Skew:                          -0.406   Prob(JB):                      0.00633
Kurtosis:                       1.670   Cond. No.                     3.14e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.14e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [58]:
#Now run the chow test which is just a specific joint hypothesis test!
hypothesisc = '(Me = 0, C0 = 0, C1 = 0, C2 = 0, C3 = 0, C4 = 0, C5 = 0, C6 = 0)'
print(Lego_modelc.wald_test(hypothesisc))

<F test: F=array([[0.80901645]]), p=0.5964160142181305, df_denom=88, df_num=8>


D:\New Program\lib\site-packages\statsmodels\base\model.py:1889: FutureWarning: The behavior of wald_test will change after 0.14 to returning scalar test statistic values. To get the future behavior now, set scalar to True. To silence this message while retaining the legacy behavior, set scalar to False.
  warnings.warn(


### Answer
P-value is at 0.5964160142181305 which is greater than alpha at 0.05. We canno reject null hypothesis. This indicate that there is no different between any of the three scenarios (i.e., me alone, my partner alone, us together)


### 2.	You are interviewing for a job at a prestigious consulting firm that focuses on analytics. The following are the questions asked by the recruiting manager.  For each question, provide a clear and concise answer that shows you understand the material from both a managerial and statistical perspective.

#### a.	 Your client hired a new consulting firm in January to adjust their pricing strategy. They would like to understand if the pricing change had an effect, and what kind of effect it had. How could you accomplish this?  3 marks

### Answer
If client would like to develop model to understand effect of the pricing change, I would recommend how to develop best model as following 
1. construct the simple model by using the theory : Testing down model like Test Test Test (TTT) theory is advised because of less bias.
2. prepare data set and seperate data set for developing, estimating, and testing.
3. use singular test of multiple constrains where possible instead of multiple test
4. check whether model violate any regression assumptions
5. understand how good fit of model (R-squared), explanatory power(F-Test), how each individual independence variable can explain   and dependence variable (T-tes), and other specification of the model  whether it is sufficient for business requirements


#### b.	You are building a model to forecast sales for your client. The project manager is insisting that all hypothesis tests have p-values be < 0.0001. What would you tell the project manager? 3 marks

### Answer 
To make a decision,  we have to compare p-value from the test with alpha. We would reject the null hypothesis and conclude alternative hypothesis if P-value is less than alpha. we mostly use significant level at 0.05 and 0.01 depend on our acceptance of type I error or how much risk we can accept for rejecting a true null. If we set alpha as 0.01, it implies that we have a low-risk tolerance on the cost of failure.

The are 2  types of test that we consider and interpret P-value. We compare p-value from the test with significant level(alpha) to interpret the result and make a decision.
1) F-Test : test explantory power of the model to the data
in this case, if p-values from F-Test are < 0.0001 means that the all tested independence variables can jointly significantly explain dependence variables which is sale in this case.

2) T-Test : test how individual variable statistically significant explain dependent variable 
in this case, if p-values from T-Test are < 0.0001 means that the tested independence variables can significantly explain dependence variables which is sale in this case.

##### c.	Your client has provided a dataset with some missing data and has told you to just impute all the missing values with the mean. What would you tell them? 3 marks

### Answer
There are 3 patterns of missing data and each pattern can be solve with different techniques. No every pattern of missing data can solve by mean substitution. Therefore, we should firstly identify pattern of missing data as following.

Here are 3 paterns of missing data 
1. Missing completely at random (MCAR) :there is no pattern of how data is missing 
2. Missing at random (MAR):  there is a pattern of how data is missing whihc can be explained by the available data. 
3. Missing not at random: there is a pattern of how data is missing whihc can be explained by other missing data. 

There are at least 3 categories of techniques can be used to deal with missing data
1. Drop observation : this tecnique only work with MCAR.If the sample is large enough that the loss of the MCAR observations from the sample have a small impact, this is certainly the easiest way to go.
2. Mean Substitution : we can substute the missing data with mean of the field.This technique is acceptable for MCAR
3. Modelling techniques : we develop model in software to estimate and assign missing data based on the remaining data